File: sample/CDMFolders/CSharp/Microsoft.CdmFolders.ObjectModel/Model/Partition.cs
// <copyright file="Partition.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// </copyright>

namespace Microsoft.CdmFolders.ObjectModel
{
    using System;
    using Microsoft.CdmFolders.ObjectModel.SerializationHelpers;
    using Newtonsoft.Json;

    /// <summary>
    /// Represents the name and location of the actual data 
    /// files corresponding to the entity definition.
    /// </summary>
    [JsonObject(MemberSerialization.OptIn)]
    public class Partition : DataObject
    {
        /// <summary>
        /// Gets or sets the Refresh Time
        /// </summary>
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        public DateTimeOffset? RefreshTime { get; set; }

        /// <summary>
        /// Gets or sets location
        /// </summary>
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        public Uri Location { get; set; }

        /// <summary>
        /// Gets or sets file format settings
        /// </summary>
        /// <remarks>The usage of <see cref="TypeNameHandling.Auto"/> is ok since the TypeNameSerializationBinder limits the scope only to this assembly</remarks>
        [JsonProperty(TypeNameHandling = TypeNameHandling.Auto, NullValueHandling = NullValueHandling.Ignore, Order = SerializationOrderConstants.ObjectSerializationOrder)]
        public FileFormatSettings FileFormatSettings { get; set; }

        /// <inheritdoc/>
        public override object Clone()
        {
            var clone = (Partition)base.Clone();
            clone.Location = (this.Location == null) ? null : new UriBuilder(this.Location).Uri;
            clone.FileFormatSettings = (FileFormatSettings)clone.FileFormatSettings?.Clone();

            return clone;
        }
    }
}

File: objectModel/Python/cdm/persistence/modeljson/types/partition.py
# ----------------------------------------------------------------------
# Copyright (c) Microsoft Corporation.
# All rights reserved.
# ----------------------------------------------------------------------

from typing import Optional, TYPE_CHECKING

from .data_object import DataObject
from .csv_format_settings import CsvFormatSettings

if TYPE_CHECKING:
    import datetime


class Partition(DataObject):
    """
    Represents the name and location of the actual data 
    files corresponding to the entity definition.
    """

    def __init__(self):
        super().__init__()

        self.refreshTime = None  # type: datetime
        self.location = ''  # type: str
        self.fileFormatSettings = None  # type: Optional[CsvFormatSettings]
        self.lastFileStatusCheckTime = None  # type: datetime
        self.lastFileModifiedTime = None  # type: datetime
